contracts/Dixel.sol

```diff
@@ -212,4 +221,105 @@ contract Dixel is Ownable, ReentrancyGuard, DixelSVGGenerator {
     function generateBase64SVG() external view returns (string memory) {
         return _generateBase64SVG(getPixelColors());
     }
+
+    /// @notice Modified from Gnosis
+    /// (https://github.com/gnosis/gp-v2-contracts/blob/main/src/contracts/libraries/GPv2SafeERC20.sol)
+    function safeTransferFrom(
```

Comment:
Is this for gas saving?

I'm not sure if it's a good idea using a custom block with an assembly implementation because I personally don't know much about assembly, so I have a zero confidence on reviewing this code block.

Do you think it is safe? Is this code block tested on production environment by any projects?

Reply:
did you check the Gnosis code?
`https://github.com/gnosis/gp-v2-contracts/blob/main/src/contracts/libraries/GPv2SafeERC20.sol`

and is this tested? YES!
In my view, this code is included in the [solmate](https://github.com/Rari-Capital/solmate), and it is also included in many projects.
and I think, from an EVM point of view, this is easier to understand!

Reply:
Yes I checked Gnosis code, but it seems their implementation is a bit different this code.
It seems it's the same code with `https://github.com/Rari-Capital/solmate/blob/main/src/utils/SafeTransferLib.sol` right? 
I will dig more into the code and test cases.

Reply:
Yes I checked the Gnosis implementation and it seems it's different from this code.

Solmate's [SafeTransferLib](https://github.com/Rari-Capital/solmate/blob/main/src/utils/SafeTransferLib.sol) seems the exactly the same as this, and I agree that it's thoroughly unit tested and audited.

However, they published [the first production release (v6)](https://github.com/Rari-Capital/solmate/pull/77) 10 days ago, so I it's probably not battle tested on production environment for a long time yet.

If the gas saving is just small (2-300 gas per call), let's skip this fancy assembly implementation for now.

Reply:
> Yes I checked the Gnosis implementation and it seems it's different from this code.

I don't understand,
The results and stack, memory produce the same results.

Reply:
SafeTransferLib.sol
```solidity
 function safeTransferFrom(
        ERC20 token,
        address from,
        address to,
        uint256 amount
    ) internal {
        bool callStatus;

        assembly {
            // Get a pointer to some free memory.
            let freeMemoryPointer := mload(0x40)

            // Write the abi-encoded calldata to memory piece by piece:
            mstore(freeMemoryPointer, 0x23b872dd00000000000000000000000000000000000000000000000000000000) // Begin with the function selector.
            mstore(add(freeMemoryPointer, 4), and(from, 0xffffffffffffffffffffffffffffffffffffffff)) // Mask and append the "from" argument.
            mstore(add(freeMemoryPointer, 36), and(to, 0xffffffffffffffffffffffffffffffffffffffff)) // Mask and append the "to" argument.
            mstore(add(freeMemoryPointer, 68), amount) // Finally append the "amount" argument. No mask as it's a full 32 byte value.

            // Call the token and store if it succeeded or not.
            // We use 100 because the calldata length is 4 + 32 * 3.
            callStatus := call(gas(), token, 0, freeMemoryPointer, 100, 0, 0)
        }

        require(didLastOptionalReturnCallSucceed(callStatus), "TRANSFER_FROM_FAILED");
    }
```

GPv2SafeERC20.sol
```solidity
    function safeTransfer(
        IERC20 token,
        address to,
        uint256 value
    ) internal {
        bytes4 selector_ = token.transfer.selector;

        // solhint-disable-next-line no-inline-assembly
        assembly {
            let freeMemoryPointer := mload(0x40)
            mstore(freeMemoryPointer, selector_)
            mstore(
                add(freeMemoryPointer, 4),
                and(to, 0xffffffffffffffffffffffffffffffffffffffff)
            )
            mstore(add(freeMemoryPointer, 36), value)

            if iszero(call(gas(), token, 0, freeMemoryPointer, 68, 0, 0)) {
                returndatacopy(0, 0, returndatasize())
                revert(0, returndatasize())
            }
        }

        require(getLastTransferResult(token), "GPv2: failed transfer");
    }
```

Oh I thought it's different because the code look a bit different.
I guess `SafeTransferLib` is the further optimized version of  `GPv2SafeERC20` but produces the same results?

Reply:
yep! Each of paired 'safeTransferFrom' and 'safeTransfer' produces the same results as the same operation.
not the `safeTransfer` and `safeTransferFrom`. I committed `transferFrom` and `transfer`, respectively.